### PR TITLE
Fix bug with DB-Forge composite foreign keys

### LINKED ISSUE #4310: Bug: DB Forge does not support adding composite foreign keys
**Direction**
It's impossible to add composite foreign key using forge

**Describe the bug**
I need to add the foreign key on multiple columns like this

```
  CONSTRAINT fk_column
     FOREIGN KEY (column1, column2)
     REFERENCES parent_table (column1, column2)
     ON UPDATE CASCADE
     ON DELETE CASCADE
```
Executing forge produces error when following is used

```
$this->forge->addForeignKey('column1,column2', 'parent_table', 'column1,column2', 'cascade', 'cascade')
```

**CodeIgniter 4 version**
dev-master

**Affected module(s)**
Database

This Issue is just to be noted. I'll work on the solution when I'm available. 

Related to:
#4120 
#4115

### PR DESCRIPTION
**Pull request by monkenWu**
**Description**
Fixes #4310 
Original PR #4931 

Change the branch base to 4.2.

I edited the `addForeignKey` method and `_processForeignKeys` method of `CodeIgniter\Database\Forge` so that they can define composite foreign keys in an array.

**Checklist:**
- [x] Securely signed commits
- [ ] Component(s) with PHPdocs
- [x] Unit testing, with >80% coverage
- [x] User guide updated
- [x] Conforms to style guide

### CODE REVIEW

**Review by michalsn — CHANGES_REQUESTED** [review]
This looks good. 

I was wondering if it would be better to leave the string casting in `addForeignKey()` method and parse the variables, but using arrays is more natural in this case. I don't think many developers have extended this method so chances that this will break someone's code are low.

I'm fine with these changes, but since we're removing casting in the public method for some variables, please mention this in the changelog (https://github.com/codeigniter4/CodeIgniter4/tree/4.2/user_guide_src/source/changelogs - you will have to create a new file `v4.2.0.rst`). We don't wanna surprise anyone.

**Comment by monkenWu** [comment]
@michalsn  Thanks for the review.
In fact, I refer to the method `addKey()`, which accepts both string and array. Therefore, I also continued this pattern in `addForeignKey()`.
擷取

I will create a new `v4.2.0.rst` file soon and mention this change in `Enhancements`.

**Comment by paulbalandan** [comment]
Please rename your last commit to something else. We're not releasing yet.

File: system/Database/Forge.php [hunk 1]
```diff
@@ -403,27 +403,38 @@ public function addField($field)
...
+        $errorNames = [];
+
+        foreach ($fieldName as $name) {
+            if (! isset($this->fields[$name])) {
+                $errorNames[] = $name;
+            }
+        }
+
+        if ($errorNames !== []) {
+            throw new DatabaseException(lang('Database.fieldNotExists', $errorNames));
```
Comment (paulbalandan) [inline system/Database/Forge.php:429]: If the count > 1, the array's first element will only be translated to the exception message. The others would be disregarded. I think you need to call implode here and adjust the message for singular and plural fields.

Comment (paulbalandan) [reply system/Database/Forge.php:429]: Please add a test showing that adding unknown fields will throw an exception.

**Comment by monkenWu** [comment]
@paulbalandan  I fixed the composition foreign keys exception and added a new test case

**Comment by lonnieezell** [comment]
@michalsn are you able to review these changes?

**Review by michalsn — APPROVED** [review]
Thanks @monkenWu it looks good!

One last thing - you have to rebase to resolve conflicts. If you have never done this before, here is a nice instruction: https://samsonasik.wordpress.com/2015/09/16/practical-git-4-rebasing-conflicted-task-branch-against-primary-branch/

File: tests/system/Database/Live/ForgeTest.php [hunk 2]
```diff
@@ -452,6 +454,135 @@ public function testForeignKeyAddingWithStringFields()
...
+
+        $this->forge->createTable('forge_test_invoices', true, $attributes);
+
+        $foreignKeyData = $this->db->getForeignKeyData('forge_test_invoices');
+
+        if ($this->db->DBDriver === 'SQLite3') {
+            $this->assertSame($foreignKeyData[0]->constraint_name, 'users_id to db_forge_test_users.id');
+            $this->assertSame($foreignKeyData[0]->sequence, 0);
+            $this->assertSame($foreignKeyData[1]->constraint_name, 'users_second_id to db_forge_test_users.second_id');
+            $this->assertSame($foreignKeyData[1]->sequence, 1);
```
Comment (paulbalandan) [inline tests/system/Database/Live/ForgeTest.php:514]: In `assertSame` the expected values should come first before the tested values.

File: tests/system/Database/Live/ForgeTest.php [hunk 3]
```diff
@@ -452,6 +454,135 @@ public function testForeignKeyAddingWithStringFields()
...
+        } else {
+            $haystack = ['users_id', 'users_second_id'];
+            $this->assertSame($foreignKeyData[0]->constraint_name, $this->db->DBPrefix . 'forge_test_invoices_users_id_users_second_id_foreign');
```
Comment (paulbalandan) [inline tests/system/Database/Live/ForgeTest.php:517]: Same

File: tests/system/Database/Live/ForgeTest.php [hunk 4]
```diff
@@ -452,6 +454,135 @@ public function testForeignKeyAddingWithStringFields()
...
+            $this->assertContains($foreignKeyData[0]->column_name, $haystack);
+
+            $secondIdKey = $this->db->DBDriver === 'Postgre' ? 2 : 1;
+            $this->assertSame($foreignKeyData[$secondIdKey]->constraint_name, $this->db->DBPrefix . 'forge_test_invoices_users_id_users_second_id_foreign');
```
Comment (paulbalandan) [inline tests/system/Database/Live/ForgeTest.php:521]: Same

File: tests/system/Database/Live/ForgeTest.php [hunk 5]
```diff
@@ -452,6 +454,135 @@ public function testForeignKeyAddingWithStringFields()
...
+            $this->assertContains($foreignKeyData[$secondIdKey]->column_name, $haystack);
+        }
+        $this->assertSame($foreignKeyData[0]->table_name, $this->db->DBPrefix . 'forge_test_invoices');
+        $this->assertSame($foreignKeyData[0]->foreign_table_name, $this->db->DBPrefix . 'forge_test_users');
+
```
Comment (paulbalandan) [inline tests/system/Database/Live/ForgeTest.php:526]: Same

**Review by paulbalandan — APPROVED** [review]

**Comment by paulbalandan** [comment]
@monkenWu can you do a last minute rebase to fix the merge conflicts before we can merge this?

**Review by paulbalandan — APPROVED** [review]

**Pull request merged**